xarray/core/parallelcompat.py

```diff
@@ -0,0 +1,199 @@
+"""
+The code in this module is an experiment in going from N=1 to N=2 parallel computing frameworks in xarray.
+It could later be used as the basis for a public interface allowing any N frameworks to interoperate with xarray,
+but for now it is just a private experiment.
+"""
+
+from abc import ABC, abstractmethod
+from typing import Any
+
+import numpy as np
+
+from . import indexing, utils
+from .pycompat import DuckArrayModule, is_duck_dask_array
+
+CHUNK_MANAGERS = {}
+
+
+def _get_chunk_manager(name: str) -> "ChunkManager":
+    if name in CHUNK_MANAGERS:
+        chunkmanager = CHUNK_MANAGERS[name]
+        return chunkmanager()
+    else:
+        raise ImportError(f"ChunkManager {name} has not been defined")
+
+
+class ChunkManager(ABC):
+    """
+    Adapter between a particular parallel computing framework and xarray.
+
+    Attributes
+    ----------
+    array_type
+        Type of the array class this parallel computing framework provides.
+
+        Parallel frameworks need to provide an array class that supports the array API standard.
+        Used for type checking.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def chunks(arr):
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def chunk(data: np.ndarray, chunks, **kwargs):
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def rechunk(data: Any, chunks, **kwargs):
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def compute(arr, **kwargs) -> np.ndarray:
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def apply_ufunc():
+        """
+        Called inside xarray.apply_ufunc, so must be supplied for vast majority of xarray computations to be supported.
+        """
+        ...
+
+    @staticmethod
+    def map_blocks():
+        """Called by xarray.map_blocks."""
+        raise NotImplementedError()
+
+    @staticmethod
+    def blockwise():
+        """Called by some niche functions in xarray."""
+        raise NotImplementedError()
```

Comment:
It would be nice if we only had to implement one of this operations, e.g., perhaps `blockwise` alone would suffice:
https://github.com/dask/dask/issues/9510

I would be curious what @tomwhite  thinks.

Reply:
It should be possible to implement all of these with `blockwise` alone, but if the chunked array library (Dask, Cubed, etc) offered implementations of the others (`apply_gufunc`, `map_blocks`), then wouldn't you want to just delegate to them? Or are there reasons for xarray to need to have its own implementation of these anyway?

Reply:
I think there's value in a minimal interface, and I doubt there's any real advatantage to using these other formulations -- other than the accident that that is how we wrote Xarray's existing implementation on top of Dask.

That said, for now it would be a great start just to get everything in one file, so the structure of our dependence on Dask becomes clear. We can save this clean-up for later.